File: python/algorithms/graph/shortest_path/bellman_ford.py
"""Bellman–Ford 单源最短路（可有负权边；可判负环）。"""

from __future__ import annotations

from typing import List, Optional, Tuple


def bellman_ford(
    n: int,
    edges: List[Tuple[int, int, int]],
    src: int,
) -> Optional[List[int]]:
    """有向边 (u, v, w)。无负环时返回 dist；从 src 可达负环则返回 None。"""
    inf = 10**18
    dist = [inf] * n
    dist[src] = 0
    for _ in range(n - 1):
        updated = False
        for u, v, w in edges:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                updated = True
        if not updated:
            break
    for u, v, w in edges:
        if dist[u] != inf and dist[u] + w < dist[v]:
            return None
    return dist


if __name__ == "__main__":
    e = [(0, 1, 4), (0, 2, 5), (1, 2, -2)]
    d = bellman_ford(3, e, 0)
    assert d == [0, 4, 2]

    neg = [(0, 1, 1), (1, 2, -1), (2, 0, -1)]
    assert bellman_ford(3, neg, 0) is None
    assert bellman_ford(1, [], 0) == [0]
    inf = 10**18
    assert bellman_ford(3, [(0, 1, 1)], 0) == [0, 1, inf]
    print("bellman_ford OK")

File: cpp/algorithms/graph/shortest_path/bellman_ford.cpp
// Bellman–Ford 单源最短路（可有负权边；可判负环）
#include <alg_std.hpp>
#include <cassert>
using namespace std;

optional<vector<long long>> bellman_ford(
    int n, const vector<array<int, 3>>& edges, int src) {
    const long long INF = (long long)4e18;
    vector<long long> dist(n, INF);
    dist[src] = 0;
    for (int i = 0; i < n - 1; ++i) {
        bool updated = false;
        for (auto [u, v, w] : edges) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                updated = true;
            }
        }
        if (!updated) break;
    }
    for (auto [u, v, w] : edges) {
        if (dist[u] != INF && dist[u] + w < dist[v]) return nullopt;
    }
    return dist;
}

int main() {
    vector<array<int, 3>> e{{0, 1, 4}, {0, 2, 5}, {1, 2, -2}};
    auto d = bellman_ford(3, e, 0);
    assert(d && *d == vector<long long>({0, 4, 2}));

    vector<array<int, 3>> neg{{0, 1, 1}, {1, 2, -1}, {2, 0, -1}};
    assert(!bellman_ford(3, neg, 0));
    cout << "bellman_ford OK" << endl;
    return 0;
}
